katdal/chunkstore_s3.py

```diff
@@ -156,7 +158,12 @@ def has_chunk(self, array_name, slices, dtype):
             return False
         else:
             actual_bytes = response['ContentLength']
-            expected_bytes = int(np.prod(shape)) * dtype.itemsize
+            header = {'shape': shape, 'fortran_order': False,
+                      'descr': np.lib.format.dtype_to_descr(dtype)}
+            fp = io.BytesIO()
+            np.lib.format._write_array_header(fp, header)
```

Comment:
I was going to suggest using write_array_header_2_0 since it's public, but that won't work because it won't necessarily match the actual header. In general, you can't be sure what the header version is used on the object without reading it (e.g. at some point numpy might change its logic that autoselects a version, or create a version 3). Does S3 have a way you can actually read back the header without reading the entire object? Otherwise your best bet might be to try version 1 and version 2 and accept it if either size matches.

Reply:
Well, the header should match if we always use v2 when writing and reading NPY. But then again, it might confuse some future (external) users who are not aware of the distinction and tries to simply `np.save` some data for reading by `katdal`. Or maybe that is improbable.

The point of `has_chunk` is to avoid reading the object, which is what `head_object` does. Most of the time the object won't be there at all - the size check is just a secondary thing *because we can*. In fact, for `S3ChunkStore.has_dask_array` I am considering checking the name only using a much faster method.

Or we could just keep using the internal numpy function and update it if numpy ever changes it (they'll surely always have a "write the header" function).

Reply:
> the size check is just a secondary thing because we can

Indeed - so whatever you choose to do, it should probably accept any valid numpy header, even if it is a different version than the one put_chunk would use in the current environment.